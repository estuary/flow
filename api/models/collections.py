from pydantic import BaseModel
from typing import Any


class CollectionSpec(BaseModel):
    name: str
    key: list[str]
    writeSchema: dict[str, Any]
    readSchema: dict[str, Any] | None = None
