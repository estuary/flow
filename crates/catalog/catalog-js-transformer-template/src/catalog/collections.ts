// This file contains placeholder examples of collection types,
// which are over-written during catalog build.

export type ExampleSourceCollection = {
    hello: string;
};

export type ExampleDerivedCollection = {
    world: string;
    value: number;
};
