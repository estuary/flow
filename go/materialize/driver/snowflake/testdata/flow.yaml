collections:
  key/value:
    schema:
      type: object
      properties:
        key1: { type: integer }
        key2: { type: boolean }
        boolean: { type: boolean }
        integer: { type: integer }
        number: { type: number }
        string: { type: string }
      required: [key1, key2]
    key: [/key1, /key2]

materializations:
  test/sqlite:
    endpoint:
      sqlite:
        path: ":memory:"
    bindings:
      - source: key/value
        resource: { table: key_value }

storageMappings:
  "": { stores: [{ provider: S3, bucket: a-bucket }] }