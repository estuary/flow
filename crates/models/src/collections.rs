use super::{CompositeKey, Derivation, Field, Id, JournalTemplate, JsonPointer, Schema};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use serde_json::{from_value, json};
use std::collections::BTreeMap;

/// Collection describes a set of related documents, where each adheres to a
/// common schema and grouping key. Collections are append-only: once a document
/// is added to a collection, it is never removed. However, it may be replaced
/// or updated (either in whole, or in part) by a future document sharing its
/// key. Each new document of a given key is "reduced" into existing documents
/// of the key. By default, this reduction is achieved by completely replacing
/// the previous document, but much richer reduction behaviors can be specified
/// through the use of annotated reduction strategies of the collection schema.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
#[schemars(example = "CollectionDef::example")]
pub struct CollectionDef {
    /// # Schema against which collection documents are validated and reduced on write and read.
    #[schemars(example = "Schema::example_relative")]
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub schema: Option<Schema>,
    /// # Schema against which collection documents are validated and reduced on write.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub write_schema: Option<Schema>,
    /// # Schema against which collection documents are validated and reduced on read.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub read_schema: Option<Schema>,
    /// # Composite key of this collection.
    pub key: CompositeKey,
    /// # Projections and logical partitions of this collection.
    #[schemars(schema_with = "projections_schema")]
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub projections: BTreeMap<Field, Projection>,
    /// # Template for journals of this collection.
    #[serde(default, skip_serializing_if = "JournalTemplate::is_empty")]
    pub journals: JournalTemplate,
    // # Derivation which builds this collection as transformations of other collections.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub derive: Option<Derivation>,
    /// # Expected publication ID of this collection within the control plane.
    /// When present, a publication of the collection will fail if the
    /// last publication ID in the control plane doesn't match this value.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expect_pub_id: Option<Id>,
    /// # Delete this collection within the control plane.
    /// When true, a publication will delete this collection.
    #[serde(default, skip_serializing_if = "super::is_false")]
    pub delete: bool,
}

impl CollectionDef {
    pub fn example() -> Self {
        Self {
            schema: Some(Schema::example_inline_basic()),
            write_schema: None,
            read_schema: None,
            key: CompositeKey::example(),
            projections: BTreeMap::new(),
            journals: JournalTemplate::default(),
            derive: None,
            expect_pub_id: None,
            delete: false,
        }
    }
}

/// Projections are named locations within a collection document which
/// may be used for logical partitioning or directly exposed to databases
/// into which collections are materialized.
#[derive(Serialize, Deserialize, Clone, Debug, JsonSchema)]
#[serde(untagged, deny_unknown_fields, rename_all = "camelCase")]
pub enum Projection {
    Pointer(JsonPointer),
    Extended {
        /// # Location of this projection.
        location: JsonPointer,
        /// # Is this projection a logical partition?
        #[serde(default)]
        partition: bool,
    },
}

impl Projection {
    pub fn as_parts(&self) -> (&JsonPointer, bool) {
        match self {
            Self::Pointer(location) => (location, false),
            Self::Extended {
                location,
                partition,
            } => (location, *partition),
        }
    }

    fn example_pointer() -> Self {
        Self::Pointer(JsonPointer::example())
    }

    fn example_extended() -> Self {
        Projection::Extended {
            location: JsonPointer::example(),
            partition: true,
        }
    }
}

fn projections_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
    let schema = Field::json_schema(gen);
    gen.definitions_mut().insert(Field::schema_name(), schema);

    let schema = Projection::json_schema(gen);
    gen.definitions_mut()
        .insert(Projection::schema_name(), schema);

    from_value(json!({
        "type": "object",
        "patternProperties": {
            Field::schema_pattern(): {
                "$ref": format!("#/definitions/{}", Projection::schema_name()),
            },
        },
        "additionalProperties": false,
        "examples": [{
            "a_field": Projection::example_pointer(),
            "a_partition": Projection::example_extended(),
        }],
    }))
    .unwrap()
}

impl super::ModelDef for CollectionDef {
    fn sources(&self) -> impl Iterator<Item = &crate::Source> {
        self.derive
            .iter()
            .map(|derive| derive.transforms.iter().map(|transform| &transform.source))
            .flatten()
    }
    fn targets(&self) -> impl Iterator<Item = &crate::Collection> {
        std::iter::empty()
    }

    fn catalog_type(&self) -> crate::CatalogType {
        crate::CatalogType::Collection
    }
}
