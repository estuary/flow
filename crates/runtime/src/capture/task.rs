use super::{Binding, Task};
use anyhow::Context;
use proto_flow::capture::{request, response, Request, Response};
use proto_flow::flow;
use std::collections::BTreeMap;

impl Task {
    pub fn new(open: &Request, opened: &Response) -> anyhow::Result<Self> {
        let request::Open {
            capture: spec,
            range,
            state_json: _,
            version: _,
        } = open.clone().open.context("expected Open")?;

        let response::Opened {
            explicit_acknowledgements,
        } = opened.clone().opened.context("expected Opened")?;

        let flow::CaptureSpec {
            bindings,
            config_json: _,
            connector_type: _,
            interval_seconds,
            name,
            network_ports: _,
            recovery_log_template: _,
            shard_template: _,
        } = spec.as_ref().context("missing capture")?;
        let range = range.context("missing range")?;

        if range.r_clock_begin != 0 || range.r_clock_end != u32::MAX {
            anyhow::bail!("captures cannot split on r-clock: {range:?}");
        }

        let ser_policy = doc::SerPolicy::default();

        let bindings = bindings
            .into_iter()
            .enumerate()
            .map(|(index, spec)| Binding::new(spec, ser_policy.clone()).context(index))
            .collect::<Result<Vec<_>, _>>()?;

        let restart = std::time::Duration::from_secs(*interval_seconds as u64);
        let restart = tokio::time::Instant::now().checked_add(restart).unwrap();

        let shard_ref = ops::ShardRef {
            kind: ops::TaskType::Capture as i32,
            name: name.clone(),
            key_begin: format!("{:08x}", range.key_begin),
            r_clock_begin: format!("{:08x}", range.r_clock_begin),
        };

        Ok(Self {
            bindings,
            explicit_acknowledgements,
            restart,
            shard_ref,
        })
    }

    pub fn binding_shapes_by_index(
        &self,
        mut by_key: BTreeMap<String, doc::Shape>,
    ) -> Vec<doc::Shape> {
        let mut by_index = Vec::new();
        by_index.resize_with(self.bindings.len(), || doc::shape::Shape::nothing());

        for (index, binding) in self.bindings.iter().enumerate() {
            let key = binding.resource_path.join("\t"); // TODO(johnny): bindings PRD.

            if let Some(shape) = by_key.remove(&key) {
                by_index[index] = shape;
            }
        }

        by_index
    }

    pub fn binding_shapes_by_key(&self, by_index: Vec<doc::Shape>) -> BTreeMap<String, doc::Shape> {
        let mut by_key = BTreeMap::new();

        for (index, shape) in by_index.into_iter().enumerate() {
            let key = self.bindings[index].resource_path.join("\t"); // TODO(johnny): bindings PRD.
            by_key.insert(key, shape);
        }
        by_key
    }

    pub fn combine_spec(&self) -> anyhow::Result<doc::combine::Spec> {
        let combiner_spec = self
            .bindings
            .iter()
            .enumerate()
            .map(|(index, binding)| binding.combiner_spec().context(index))
            .collect::<Result<Vec<_>, _>>()?;

        let state_schema = doc::reduce::merge_patch_schema().to_string();
        let state_schema = doc::validation::build_bundle(&state_schema).unwrap();
        let state_validator = doc::Validator::new(state_schema).unwrap();

        // Build combiner Spec with all bindings, plus one extra for state reductions.
        let combiner_spec = doc::combine::Spec::with_bindings(
            combiner_spec
                .into_iter()
                .map(|(is_full, key, validator)| (is_full, key, None, validator))
                .chain(std::iter::once((false, Vec::new(), None, state_validator))),
        );

        Ok(combiner_spec)
    }
}

impl Binding {
    pub fn new(
        spec: &flow::capture_spec::Binding,
        ser_policy: doc::SerPolicy,
    ) -> anyhow::Result<Self> {
        let flow::capture_spec::Binding {
            backfill: _,
            collection,
            resource_config_json: _,
            resource_path,
            state_key: _,
        } = spec;

        let flow::CollectionSpec {
            ack_template_json: _,
            derivation: _,
            key,
            name,
            partition_fields,
            partition_template: _,
            projections,
            read_schema_json: _,
            uuid_ptr,
            write_schema_json,
        } = collection.as_ref().context("missing collection")?;

        let document_uuid_ptr = doc::Pointer::from(uuid_ptr);
        let key_extractors = extractors::for_key(&key, &projections, &ser_policy)?;
        let partition_extractors =
            extractors::for_fields(&partition_fields, &projections, &ser_policy)?;

        Ok(Self {
            collection_name: name.clone(),
            document_uuid_ptr,
            key_extractors,
            partition_extractors,
            resource_path: resource_path.clone(),
            ser_policy,
            write_schema_json: write_schema_json.clone(),
        })
    }

    pub fn combiner_spec(&self) -> anyhow::Result<(bool, Vec<doc::Extractor>, doc::Validator)> {
        let built_schema = doc::validation::build_bundle(&self.write_schema_json)
            .context("collection write_schema_json is not a JSON schema")?;
        let validator =
            doc::Validator::new(built_schema).context("could not build a schema validator")?;

        Ok((false, self.key_extractors.clone(), validator))
    }
}
