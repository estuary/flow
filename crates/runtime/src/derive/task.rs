use super::{Task, Transform};
use anyhow::Context;
use proto_flow::derive::{request, response, Request, Response};
use proto_flow::flow;

impl Task {
    pub fn new(open: &Request, opened: &Response) -> anyhow::Result<Self> {
        let request::Open {
            collection,
            range,
            state_json: _,
            version: _,
        } = open.clone().open.context("expected Open")?;

        let response::Opened {} = opened.opened.as_ref().context("expected Opened")?;

        let flow::CollectionSpec {
            ack_template_json: _,
            derivation,
            key,
            name: collection_name,
            partition_fields,
            partition_template: _,
            projections,
            read_schema_json: _,
            uuid_ptr,
            write_schema_json,
        } = collection.context("missing collection")?;

        let flow::collection_spec::Derivation {
            config_json: _,
            connector_type: _,
            network_ports: _,
            recovery_log_template: _,
            shard_template: _,
            shuffle_key_types: _,
            transforms,
        } = derivation.as_ref().context("missing derivation")?;

        if key.is_empty() {
            anyhow::bail!("collection key cannot be empty");
        }

        let range = range.context("missing range")?;
        let ser_policy = doc::SerPolicy::default();

        let document_uuid_ptr = doc::Pointer::from(uuid_ptr);
        let key_extractors = extractors::for_key(&key, &projections, &ser_policy)?;
        let partition_extractors =
            extractors::for_fields(&partition_fields, &projections, &ser_policy)?;

        let shard_ref = ops::ShardRef {
            kind: ops::TaskType::Derivation as i32,
            name: collection_name.clone(),
            key_begin: format!("{:08x}", range.key_begin),
            r_clock_begin: format!("{:08x}", range.r_clock_begin),
        };

        let transforms = transforms
            .into_iter()
            .enumerate()
            .map(|(index, spec)| Transform::new(spec).context(index))
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self {
            collection_name,
            document_uuid_ptr,
            key_extractors,
            partition_extractors,
            ser_policy,
            shard_ref,
            transforms,
            write_schema_json,
        })
    }

    pub fn combine_spec(&self) -> anyhow::Result<doc::combine::Spec> {
        let built_schema = doc::validation::build_bundle(&self.write_schema_json)
            .context("collection write_schema_json is not a JSON schema")?;
        let validator =
            doc::Validator::new(built_schema).context("could not build a schema validator")?;

        Ok(doc::combine::Spec::with_one_binding(
            false,
            self.key_extractors.clone(),
            None,
            validator,
        ))
    }

    pub fn validators(&self) -> anyhow::Result<Vec<doc::Validator>> {
        self.transforms
            .iter()
            .map(|transform| {
                transform.validator().with_context(|| {
                    format!("failed to build validator for transform {}", transform.name)
                })
            })
            .collect::<Result<Vec<_>, _>>()
    }
}

impl Transform {
    pub fn new(spec: &flow::collection_spec::derivation::Transform) -> anyhow::Result<Self> {
        let flow::collection_spec::derivation::Transform {
            backfill: _,
            collection,
            journal_read_suffix: _,
            lambda_config_json: _,
            name,
            not_after: _,
            not_before: _,
            partition_selector: _,
            priority: _,
            read_delay_seconds: _,
            read_only: _,
            shuffle_key: _,
            shuffle_lambda_config_json: _,
        } = spec;

        let flow::CollectionSpec {
            ack_template_json: _,
            derivation: _,
            key: _,
            name: collection_name,
            partition_fields: _,
            partition_template: _,
            projections: _,
            read_schema_json,
            uuid_ptr: _,
            write_schema_json,
        } = collection.as_ref().context("missing collection")?;

        let read_schema_json = if read_schema_json.is_empty() {
            write_schema_json
        } else {
            read_schema_json
        }
        .clone();

        Ok(Self {
            collection_name: collection_name.clone(),
            name: name.clone(),
            read_schema_json,
        })
    }

    pub fn validator(&self) -> anyhow::Result<doc::Validator> {
        let built_schema = doc::validation::build_bundle(&self.read_schema_json)
            .context("collection read_schema_json is not a JSON schema")?;
        let validator =
            doc::Validator::new(built_schema).context("could not build a schema validator")?;
        Ok(validator)
    }
}
