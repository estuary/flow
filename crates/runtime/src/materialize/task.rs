use super::{Binding, Task};
use anyhow::Context;
use proto_flow::flow;
use proto_flow::materialize::{request, Request};

impl Task {
    pub fn new(open: &Request) -> anyhow::Result<Self> {
        let request::Open {
            materialization: spec,
            range,
            state_json: _,
            version: _,
        } = open.clone().open.context("expected Open")?;

        let flow::MaterializationSpec {
            bindings,
            config_json,
            connector_type: _,
            name,
            network_ports: _,
            recovery_log_template: _,
            shard_template: _,
        } = spec.as_ref().context("missing materialization")?;
        let range = range.context("missing range")?;

        if range.r_clock_begin != 0 || range.r_clock_end != u32::MAX {
            anyhow::bail!("materialization cannot split on r-clock: {range:?}");
        }

        // TODO(johnny): Hack to limit serialized value sizes for these common materialization connectors
        // that don't handle large strings very well. This should be negotiated via connector protocol.
        // See go/runtime/materialize.go:135
        let ser_policy = if [
            "ghcr.io/estuary/materialize-snowflake",
            "ghcr.io/estuary/materialize-redshift",
            "ghcr.io/estuary/materialize-sqlite",
        ]
        .iter()
        .any(|image| config_json.contains(image))
        {
            doc::SerPolicy {
                str_truncate_after: 1 << 16, // Truncate at 64KB.
                nested_obj_truncate_after: 1000,
                array_truncate_after: 1000,
                ..doc::SerPolicy::noop()
            }
        } else {
            doc::SerPolicy::noop()
        };

        let bindings = bindings
            .into_iter()
            .enumerate()
            .map(|(index, spec)| Binding::new(spec, &ser_policy).context(index))
            .collect::<Result<Vec<_>, _>>()?;

        let shard_ref = ops::ShardRef {
            kind: ops::TaskType::Materialization as i32,
            name: name.clone(),
            key_begin: format!("{:08x}", range.key_begin),
            r_clock_begin: format!("{:08x}", range.r_clock_begin),
        };

        Ok(Self {
            bindings,
            shard_ref,
        })
    }

    pub fn combine_spec(&self) -> anyhow::Result<doc::combine::Spec> {
        let combiner_spec = self
            .bindings
            .iter()
            .enumerate()
            .map(|(index, binding)| binding.combiner_spec().context(index))
            .collect::<Result<Vec<_>, _>>()?;

        // Build combiner Spec with all bindings, plus one extra for state reductions.
        let combiner_spec = doc::combine::Spec::with_bindings(
            combiner_spec
                .into_iter()
                .map(|(is_full, key, name, validator)| (is_full, key, name, None, validator)),
        );

        Ok(combiner_spec)
    }
}

impl Binding {
    pub fn new(
        spec: &flow::materialization_spec::Binding,
        ser_policy: &doc::SerPolicy,
    ) -> anyhow::Result<Self> {
        let flow::materialization_spec::Binding {
            backfill: _,
            collection,
            delta_updates,
            deprecated_shuffle: _,
            field_selection,
            journal_read_suffix,
            not_after: _,
            not_before: _,
            partition_selector: _,
            priority: _,
            resource_config_json: _,
            resource_path: _,
            state_key,
        } = spec;

        let flow::FieldSelection {
            document: selected_root,
            field_config_json_map: _,
            keys: selected_key,
            values: selected_values,
        } = field_selection
            .as_ref()
            .context("missing field selection")?;

        let flow::CollectionSpec {
            ack_template_json: _,
            derivation: _,
            key: _,
            name: collection_name,
            partition_fields: _,
            partition_template: _,
            projections,
            read_schema_json,
            uuid_ptr: _,
            write_schema_json,
        } = collection.as_ref().context("missing collection")?;

        let key_extractors = extractors::for_fields(selected_key, projections, ser_policy)?;
        let value_extractors = extractors::for_fields(selected_values, projections, ser_policy)?;

        let read_schema_json = if read_schema_json.is_empty() {
            write_schema_json
        } else {
            read_schema_json
        }
        .clone();

        Ok(Self {
            collection_name: collection_name.clone(),
            delta_updates: *delta_updates,
            journal_read_suffix: journal_read_suffix.clone(),
            key_extractors,
            read_schema_json,
            ser_policy: ser_policy.clone(),
            state_key: state_key.clone(),
            store_document: !selected_root.is_empty(),
            value_extractors,
        })
    }

    pub fn combiner_spec(
        &self,
    ) -> anyhow::Result<(bool, Vec<doc::Extractor>, String, doc::Validator)> {
        let built_schema = doc::validation::build_bundle(&self.read_schema_json)
            .context("collection read_schema_json is not a JSON schema")?;
        let validator =
            doc::Validator::new(built_schema).context("could not build a schema validator")?;

        Ok((
            !self.delta_updates,
            self.key_extractors.clone(),
            format!("materialized collection {}", self.collection_name),
            validator,
        ))
    }
}
