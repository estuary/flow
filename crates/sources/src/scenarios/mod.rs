use crate::{Fetcher, Loader, Scope};
use futures::channel::oneshot;
use futures::future::{FutureExt, LocalBoxFuture};
use proto_flow::flow;
use std::cell::RefCell;
use std::collections::BTreeMap;
use std::task::Poll;
use url::Url;

#[cfg(test)]
mod test {
    use super::evaluate_fixtures;

    macro_rules! file_tests {
    ($($name:ident,)*) => {
    $(
        #[test]
        fn $name() {
			let fixture = include_bytes!(concat!(stringify!($name), ".yaml"));
            let fixture: serde_json::Value = serde_yaml::from_slice(fixture).unwrap();
			let mut tables = evaluate_fixtures(Default::default(), &fixture);
			insta::assert_debug_snapshot!(tables);

            crate::inline_sources(&mut tables);

            // Clear out load-related state prior to the snapshot.
            tables.errors = tables::Errors::new();
            tables.fetches = tables::Fetches::new();
            tables.resources = tables::Resources::new();

            // Verify shape of inline specs.
			insta::assert_debug_snapshot!(tables);

            // Now indirect specs again, and verify the updated specs and indirect'd resources.
            crate::indirect_large_files(&mut tables, 32);
            crate::rebuild_catalog_resources(&mut tables);
			insta::assert_debug_snapshot!(tables);
        }
    )*
    }
}
    file_tests! {
        test_catalog_import_cycles,
        test_collections,
        test_derivations,
        test_endpoints_captures_materializations,
        test_schema_with_anchors,
        test_schema_with_inline,
        test_schema_with_nested_ids,
        test_schema_with_references,
        test_simple_catalog,
        test_storage_mappings,
        test_test_case,
    }
}
// MockFetcher queues and returns oneshot futures for started fetches.
struct MockFetcher<'f> {
    fetches: &'f RefCell<BTreeMap<String, oneshot::Sender<Result<bytes::Bytes, anyhow::Error>>>>,
}

impl<'f> Fetcher for MockFetcher<'f> {
    fn fetch<'a>(
        &self,
        resource: &'a Url,
        _content_type: flow::ContentType,
    ) -> LocalBoxFuture<'a, Result<bytes::Bytes, anyhow::Error>> {
        let (tx, rx) = oneshot::channel();

        if let Some(_) = self.fetches.borrow_mut().insert(resource.to_string(), tx) {
            panic!("resource {} has already been fetched", resource);
        }
        rx.map(|r| r.unwrap()).boxed_local()
    }
}

pub fn evaluate_fixtures(catalog: tables::Sources, fixture: &serde_json::Value) -> tables::Sources {
    let fixtures = match fixture {
        serde_json::Value::Object(m) => m,
        _ => panic!("fixtures must be an object having resource properties"),
    };

    // Fetches holds started fetches since the last future poll.
    // Use an ordered map so that we signal one-shots in a stable order,
    // making snapshots reliable.
    let fetches = RefCell::new(BTreeMap::new());

    let loader = Loader::new(catalog, MockFetcher { fetches: &fetches });
    let root = Url::parse("test://example/catalog.yaml").unwrap();

    // What's going on here? Glad you asked.
    //
    // loader.load_resource() is returning a Future, which under the covers
    // is a big, synchronously-invoked, and deterministic state machine that's
    // built with the compiler's assistance.
    //
    // By construction, we know that there's only one real "await" point of this
    // future which will cause it to return Poll::Pending: when it's called into the
    // |fetch| closure we gave it above, and it has no work remaining to do until
    // at least one of the Futures returned by |fetch| resolves.
    //
    // Loader is walking sources concurrently. It processes fetches as separate
    // internal (still synchronous & deterministic) tasks, so we *do* expect to
    // see multiple |fetch| calls made in between Poll::Pending poll results of
    // the future.
    //
    // So, the strategy is this: we _synchronously_ poll the future forward in a
    // loop (e.x., we're not using a runtime *at all*). Every time it returns
    // Poll::Pending, we assert that it's queued new calls to |fetch| during this
    // iteration, and we resolve each of those response futures. Eventually, it
    // returns Poll::Ready when it's fully walked the source fixture, and at that
    // point we unwrap and return the loaded Tables.
    //
    // Note that the use of BTreeMap above is significant: it means that we resolve
    // those call futures in a stable (sorted) order, and that's the order with which
    // the future will process resolved responses on it's next poll(). It makes the
    // whole mess fully deterministic.

    let mut fut = loader
        .load_resource(Scope::new(&root), &root, flow::ContentType::Catalog)
        .boxed_local();

    let waker = futures::task::noop_waker();
    let mut ctx = std::task::Context::from_waker(&waker);

    loop {
        match fut.poll_unpin(&mut ctx) {
            Poll::Ready(()) => {
                std::mem::drop(fut);
                return loader.into_tables();
            }
            Poll::Pending if fetches.borrow().is_empty() => {
                // Note the future can return Pending *only because* it's blocked
                // waiting for one or more |fetch| fixtures above to resolve.
                panic!("future is pending, but started no fetches")
            }
            Poll::Pending => {
                for (url, tx) in fetches.borrow_mut().split_off("") {
                    match fixtures.get(&url) {
                        Some(serde_json::Value::String(value)) => tx.send(Ok(value.clone().into())),
                        Some(value) => tx.send(Ok(serde_json::to_vec(&value).unwrap().into())),
                        None => tx.send(Err(anyhow::anyhow!("fixture not found"))),
                    }
                    .unwrap();
                }
            }
        }
    }
}
