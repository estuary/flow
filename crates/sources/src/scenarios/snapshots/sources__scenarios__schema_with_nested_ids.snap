---
source: crates/sources/src/scenarios/mod.rs
expression: tables

---
Tables {
    capture_bindings: [],
    captures: [],
    collections: [
        Collection {
            scope: test://example/catalog.yaml#/collections/a~1collection,
            collection: a/collection,
            schema: test://example/actual,
            key: ["/key"],
        },
    ],
    derivations: [],
    errors: [],
    fetches: [
        Fetch {
            resource: test://example/actual,
        },
        Fetch {
            resource: test://example/catalog.ts,
        },
        Fetch {
            resource: test://example/catalog.yaml,
        },
    ],
    imports: [
        Import {
            scope: test://example/catalog.yaml#/collections/a~1collection/schema,
            from_resource: test://example/catalog.yaml,
            to_resource: test://example/actual,
        },
        Import {
            scope: test://example/catalog.yaml,
            from_resource: test://example/catalog.yaml,
            to_resource: test://example/catalog.ts,
        },
    ],
    journal_rules: [],
    materialization_bindings: [],
    materializations: [],
    named_schemas: [],
    npm_dependencies: [],
    projections: [],
    resources: [
        Resource {
            resource: test://example/actual,
            content_type: "JsonSchema",
            content: ".. binary ..",
        },
        Resource {
            resource: test://example/catalog.yaml,
            content_type: "CatalogSpec",
            content: ".. binary ..",
        },
    ],
    schema_docs: [
        SchemaDoc {
            schema: test://example/actual,
            dom: {"$defs":{"wrapper":{"$defs":{"a":{"$id":"test://fake/other/a-doc","items":[true,{"$ref":"b-doc#/items/1"}]},"b":{"$id":"test://fake/other/b-doc","items":[{"$ref":"a-doc#/items/0"},true]},"c":true},"$id":"test://fake/root","allOf":[{"$ref":"other/a-doc#/items/1"},{"$ref":"test://fake/other/b-doc#/items/0"},{"$ref":"#/$defs/c"},{"$ref":"test://fake/root#/$defs/c"}]}},"$ref":"test://fake/root"},
        },
    ],
    shard_rules: [],
    test_steps: [],
    transforms: [],
}
